### DISCUSSION TITLE
Add decorators for eloquent models

### DISCUSSION THREAD

**Discussion by diegogit03** [laravel/framework#50441] [Ideas]
I thinking on this:

```php
class User extends Model {
    #[Column]
    public int $id;

    #[Column]
    public string $name;

    #[HasOne(Profile::class)]
    public Profile $profile;
}
```

Motivation:

It's more elegant and i think that can help with intelisense but it's not confirmed yet.

Problems:

- Maybe a breaking changes (i dont deep in the code yet)

References:

- [Doctrine ORM](https://www.doctrine-project.org/projects/doctrine-orm/en/3.1/reference/basic-mapping.html)
- [Adonisjs Lucid](https://lucid.adonisjs.com/docs/models#creating-your-first-model)

**Comment by ghost** [comment]
Hi. This should NOT be the default. 

```
public function relation(): HasMany | HasManyThrough
{
   if ($this->exists() && $this->parent_id !== null) {
      return $this->hasManyThrough(.....
    }
    return $this->hasMany(....
}


Or


return $this->hasMany(...)->where('column', 'text');
```

The Attributes are restricting the functionality and generating parallel code interpreting.... bad Idea from my point of view.
If Laravel goes on the road of symfony with these annotated Attributes, I will be forced to fork an older version of laravel and maintain it my self....


Analog for declaring properties in the model. OOP exageration that generates waste of money on codding... Symfony way again...

**Reply by diegogit03** [reply]
I think that laravel can support both approaches i dont know why not

**Reply by ghost** [reply]
I am curious. Why people like to complicate things. Laravel is php. Why introduce this pseudo language for Attributes (this is a question also for php itself since it introduced these attributes)? Symfony has config in yml and these attributes already.... So now instead of doing just php, you must do also yml, also abstract attributes language. Also the opcache must be set to compile also comments because they are needed... mind blowing... What happened to KISS principle?

I would like a framework that sits down and stops introducing new features that are NOT USED BY THE MAJORITY of people anyway. We should pick a version that has the **least bugs or any** and maintain it from the security point of view and from the php version point of view. And maybe once every 5 years release another LTS with new features based on votes from the users (they should be tens of millions votes for them to count for something).

**Reply by henzeb** [reply]
I agree with @marius-mcp, ever since the introduction of attributes I am not a big fan of them...

**Reply by shaedrich** [reply]
@henzeb The big problem of the attributes compared to [JavaScript's decorators](https://github.com/tc39/proposal-decorators) or [Rust decorators](https://docs.rs/decorators/latest/decorators/) is, that they don't execute any logic on their own, which makes them way less useful as is.
@marius-mcp What Laravel does, is also "magic", abstracting class properties into getters and setters and what not. So, if you're so strictly against that, maybe, Laravel isn't the best for you as a whole either. Maybe, vanilla PHP it is, then. Oh wait, attributes _are_ vanilla PHP—more than Laravel is, actually—I have a hunch, PHP as a language might not be suitable for you, then 🤔

**Reply by diegogit03** [reply]
Dude, why pseudo lang? It's pure PHP and Laravel is based on symphony components and I'm not talking about presenting a yml or json file

**Reply by Sophist-UK** [reply]
> @shaedrich + @henzeb + @marius-mcp The big problem of the attributes compared to [JavaScript's decorators](https://github.com/tc39/proposal-decorators) or [Rust decorators](https://docs.rs/decorators/latest/decorators/) is, that they don't execute any logic on their own, which makes them way less useful as is. 

Or indeed [Python decorators](https://peps.python.org/pep-0318/).

I don't know much about decorators in JS or Rust, but I am definitely in two minds about Python decorators. OTOH they simplify code by describing (and altering) the way the method they apply to works, but OTOH I personally find them a tad difficult to get to grips with understanding other people's decorators or to write my own - but on balance I think they are a powerful and useful addition.

Python Decorators are an implementation of the [Decorator design-pattern](https://refactoring.guru/design-patterns/decorator/) using Attributes as a syntax mechanism - like JS or Rust they actually generate additional executable code.

By comparison, [PHP Attributes](https://www.php.net/manual/en/language.attributes.overview.php) - which were only first delivered in PHP 8.0 - cannot themselves generate executable code, only provide metadata that can be obtained via the PHP Reflection APIs and then used by existing code to modify their behaviour. Had PHP Attributes been available when Laravel was first being developed, Taylor might well have delivered some functionality using them, however since they were effectively only available to Taylor to use with [Laravel 9](https://laravel.com/docs/9.x/releases#support-policy) - the first release that supported only Laravel 8.0 or later - backwards compatibility was/is essential, so implementing code in Laravel to use attributes would have to be a parallel alternative to other existing approaches and this would complicate the code substantially. Taylor is understandably reluctant to add complexity and code maintenance overhead to Laravel without there being substantial benefits, and parallel code wouldn't seem to meet that criterion.

So despite being generally in favour of Python Decorators implemented as Attributes, I agree with others here that implementing pseudo-decorators using Attributes is not a sensible thing to do **unless** it is part of a long-term (say 5-10 year) strategy to switch completely to Attributes as pseudo-decorators over (say) 3-5 versions of Laravel in an approach similar to the way that PHP has deprecated and migrated its legacy code-base and that the benefits of this move are so substantial as to justify the pain and effort that it would cause.

**Reply by shaedrich** [reply]
> Or indeed [Python decorators](https://peps.python.org/pep-0318/).

Thanks for that hint—very interesting read 👍🏻 

> Had PHP Attributes been available when Laravel was first being developed, Taylor might well have delivered some functionality using them

Yeah, I think, they are a huge miss. But there's a good chance, they are widely used in Laravel packages. An advantage of using them with Laravel is, that via the service container, they actually can have executable functionality on the framework level.

**Reply by ghost** [reply]
@shaedrich  Anything is possible. I'm not defining my self as a developer but as an engineer. I learned to code because I was not pleased by how developers see things in comparison with how a client sees them (I was PM for an end client, so from there the experience not 100% ok with developers).

**Reply by ghost** [reply]
@diegogit03 maybe my terminology is incorrect but when you rely on reflection to read the Attributes... to me does not seem to be a pure php code...

Also symfony way is not the best way. All frameworks that take more time do deliver mean more money milked from the clients. If it were your money on the line, you would think more before choosing symfony over laravel for example.

**Reply by shaedrich** [reply]
> not […] pure php code

@marius-mcp Sure, it is. What else should it be? Reflection is done within PHP.

> Also symfony way is not the best way. All frameworks that take more time do deliver mean more money milked form the clients. If it were your money on the line, you would think more before choosing symfony over laravel for example.

Laravel is based on Symfony. So, it's no surprise, that it has more features. I don't think, money is the only and defining factor. Taste is another one, I guess.

> I'm not defining my self as a developer but as an engineer […] so from there the experience not 100% ok with developers

Yeah, web application development is more than writing code. Just getting stuff done can result in a multitude of different codes, but they all get the job done for the time being. However, deciding, which code does it in the most efficient, future-proof way, is the hard part and even there, hardly everyone agreed on one solution only either.

**Reply by Sophist-UK** [reply]
@marius-mcp
I am with you regarding the need for developers to do what users want rather than what they think is right - or in other words "usability" rather than "technical perfection".

That said, reading attributes via reflection is IMO pure PHP. You could achieve the same functionality either by using code attributes or by requiring a method to be called to set the same information as a model attribute. The only difference is that code attribute values are essentially constants, whilst mode attribute values set via a method can be dynamic.

**Reply by shaedrich** [reply]
> I am with you regarding the need for developers to do what users want rather than what they think is right - or in other words "usability" rather than "technical perfection".

@marius-mcp @Sophist-UK However, users need to know and be able to articulate what they want. If that can't be properly communicated, technical debt and lots of adjustments are as inevitable as when developers have their way.

> The only difference is that code attribute values are essentially constants, whilst mode attribute values set via a method can be dynamic.

When dynamic and more complex constructs are not needed, static/constant expressions are usually preferred, because they don't need to be calculated at runtime, something, static code analysis can benefit from. Also, especially, the way, attributes are written, their naming could result in self-explanatory annotations to the method that can abstract away implementation details the developer doesn't necessarily need to know or worry about.

**Reply by Sophist-UK** [reply]
> When dynamic and more complex constructs are not needed, static/constant expressions are usually preferred, because they don't need to be calculated at runtime, something, static code analysis can benefit from. Also, especially, the way, attributes are written, their naming could result in self-explanatory annotations to the method that can abstract away implementation details the developer doesn't necessarily need to know or worry about.

Actually, whilst this is true in the most general and abstract case, in this specific instance I would argue the opposite is true. Using PHP attributes, you would either need to call the reflection API on every lookup or cache the values, and so at best this is going to be no faster than setting the values once with a method call, and at worse significantly worse.

**Reply by shaedrich** [reply]
@Sophist-UK There's the danger of that, yes. But this way, the static/constant expression style would be required when used this way, so that would be the only way (if not used dynamically) to implement it and the part, the user has to take care of. As for the rest, we should assume, the one writing the package is (hopefully) more experienced than the average developer using it, so they should take care of caching and efficient code. So, ideally, this should be slightly more fool-proof.

**Reply by Sophist-UK** [reply]
In the end it doesn't matter. I can't see Taylor wanting to support two alternative approaches forever, so unless there is an exceptionally good reason to switch to PHP attributes from current methods, so good that forcing users to put all existing code-bases through a migration (even if over several major versions of Laravel), then this debate is a bit pointless.

However, I think that if you can make the case to the community that this pain is worthwhile and get some good degree of community consensus that this is so, then he might well consider it.

**Reply by henzeb** [reply]
Indeed.in the mean time, feel free to create a package, Taylor would say ...

**Reply by shaedrich** [reply]
> However, I think that if you can make the case to the community that this pain is worthwhile and get some good degree of community consensus that this is so, then he might well consider it.

Oh, I think, unless you are a friend of Taylor, hoping to get a PR approved, is rather pointless.

**Comment by shaedrich** [comment]
I used a similar approach, however, not by annotating properties but the model itself, since the columns are technically not on the model but on `$this->attributes`.

**Comment by smares** [comment]
> I think that laravel can support both approaches i dont know why not

Because it's a PITA as developer having to look in multiple places what is happening and Laravel already includes a lot of "magic".

**Reply by diegogit03** [reply]
What is PITA?

Sorry I am a beginner English speaker

**Reply by Sophist-UK** [reply]
***P***ain ***I***n ***T***he ***A***ss.

**Reply by diegogit03** [reply]
LOL

**Comment by diegogit03** [comment]
Okay guys i read all replies and

Possible problems of this implementation:

- Migration of code base (but i think that we can TRY keep all compatible)
- Add more complexity (i dont agree with this)

Possible motivations:

- More elegant syntax (relative)
- Possible intelisense improvements (possibility)

With this conclusions the best solution FOR NOW is **make a package** . In my opinion, if the generic [RFC](https://wiki.php.net/rfc/generics) passes, we could go deeper into this discussion again.

**Reply by smares** [reply]
That RFC is 8 years old and originally targeted PHP 7.1. Any specific reason why you mention it?

**Reply by diegogit03** [reply]
With generics decorators approach can make more sense, like adonisjs make(it's typescript)

**Reply by ghost** [reply]
Hi. That RFC was dis-considered anyway because it is easier to write like this:

```
class A
{
    public function __construct(protected string $x, protected Carbon $c)
    {
    }
}

$a = new A('aaa', new Carbon())
```

vs

  ```
class Entry<KeyType, ValueType>
  {
      protected $key;
      protected $value;
   
      public function __construct(KeyType $key, ValueType $value)
      {
          $this->key  = $key;
          $this->value = $value;
      }
  }

$entry = new Entry<int,string>(1, 'test');
```
The only thing kept from that RFS was:

```
/**
 * @param Collection<Model> $models
 */
pubic function x(Collection $models)
```
Which I find useful.
